modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java

```diff
@@ -0,0 +1,330 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.util.IgniteUtils.nl;
+
+/** */
+public class ReentrantReadWriteLockWithTracking implements ReadWriteLock {
+    /** Lock hold message. */
+    public static final String LOCK_HOLD_MESSAGE = "ReadLock held the lock more than ";
+
+    /** Lock print threshold. */
+    private long readLockThreshold;
+
+    /** Delegate instance. */
+    private final ReentrantReadWriteLock delegate = new ReentrantReadWriteLock();
+
+    /** Read lock holder. */
+    private ReentrantReadWriteLockWithTracking.ReadLock readLock;
+
+    /** Write lock holder. */
+    private ReentrantReadWriteLockWithTracking.WriteLock writeLock = new ReentrantReadWriteLockWithTracking.WriteLock(delegate);
+
+    /**
+     * ReentrantRWLock wrapper, provides additional trace info on {@link ReadLockWithTracking#unlock()} method, if someone
+     * holds the lock more than {@code readLockThreshold}.
+     *
+     * @param log Ignite logger.
+     * @param readLockThreshold ReadLock threshold timeout.
+     */
+    public ReentrantReadWriteLockWithTracking(IgniteLogger log, long readLockThreshold) {
+        readLock = new ReadLockWithTracking(delegate, log, readLockThreshold);
+
+        this.readLockThreshold = readLockThreshold;
+    }
+
+    /** Delegator implementation. */
+    public ReentrantReadWriteLockWithTracking() {
+        readLock = new ReentrantReadWriteLockWithTracking.ReadLock(delegate);
+    }
+
+    /** {@inheritDoc} */
+    @Override public ReentrantReadWriteLockWithTracking.ReadLock readLock() {
```

Comment:
Do you really need ReentrantReadWriteLockWithTracking.ReadLock here? Maybe it makes sense to change it to java.util.concurrent.locks.ReentrantReadWriteLock.ReadLock?